seat.c

```diff
@@ -273,21 +270,170 @@ handle_keyboard_key(struct wl_listener *listener, void *data)
 	wlr_idle_notify_activity(seat->server->idle, seat->seat);
 }
 
+static void
+cg_keyboard_group_remove(struct cg_keyboard *keyboard);
+
+static void
+cg_keyboard_destroy(struct cg_keyboard *keyboard)
+{
+	struct cg_seat *seat = keyboard->seat;
+
+	if (keyboard->device->keyboard->group != NULL) {
+		cg_keyboard_group_remove(keyboard);
+	}
+
+	if (wlr_seat_get_keyboard(seat->seat) == keyboard->device->keyboard) {
+		wlr_seat_set_keyboard(seat->seat, NULL);
+	}
+
+	free(keyboard);
+}
+
+static void
+destroy_empty_wlr_keyboard_group(void *data)
+{
+	wlr_keyboard_group_destroy(data);
+}
+
+static void
+cg_keyboard_group_remove(struct cg_keyboard *keyboard)
+{
+	struct wlr_keyboard_group *wlr_group = keyboard->device->keyboard->group;
+
+	wlr_keyboard_group_remove_keyboard(keyboard->device->keyboard->group,
+	                                   keyboard->device->keyboard);
+
+	if (wl_list_empty(&wlr_group->devices)) {
+		wlr_log(WLR_DEBUG, "Destroying empty keyboard group %p", (void*)wlr_group);
+		struct cg_keyboard_group *cg_group = wlr_group->data;
+		wlr_group->data = NULL;
+		wl_list_remove(&cg_group->link);
+		wl_list_remove(&cg_group->key.link);
+		wl_list_remove(&cg_group->modifiers.link);
+		cg_keyboard_destroy(cg_group->keyboard);
```

Comment:
This can be `cb_keyboard_destroy(keyboard);`

Reply:
Ah, no this is not actually the same. `cg_group->keyboard` is a reference to a kind of virtual keyboard which consists of the sum of all the keyboards in the group, whereas `keyboard` is the specific (physical) keyboard that we are removing.

Reply:
Interesting! Thanks for the explanation.